### PR TITLE
eglstream-kms: Make It Work Again™

### PR DESCRIPTION
**Pull request by RAOF**

### CODE REVIEW

**Comment by RAOF** [comment]
@gerboland Hey, yo! Give this a whirl, will you? You'll also need #414 for probing to work correctly.

**Comment by gerboland** [comment]
Sorry, am missing the pings somehow

Running this I get

> sudo XDG_RUNTIME_DIR=/tmp  ./bin/miral-shell --vt 1
> MIR_CLIENT_PLATFORM_PATH=./bin/../lib/client-modules/
> MIR_SERVER_PLATFORM_PATH=./bin/../lib/server-modules/
> LD_LIBRARY_PATH=./bin/../lib
> exec=./bin/miral-shell.bin
> [2018-06-19 18:07:52.677836] mirserver: Starting
> [2018-06-19 18:07:52.678029] mircommon: Loading modules from: ./bin/../lib/server-modules/
> [2018-06-19 18:07:52.678073] mircommon: Loading module: ./bin/../lib/server-modules/server-mesa-x11.so.15
> [2018-06-19 18:07:52.678082] mircommon: Loading module: ./bin/../lib/server-modules/graphics-mesa-kms.so.15
> [2018-06-19 18:07:52.678088] mircommon: Loading module: ./bin/../lib/server-modules/graphics-eglstream-kms.so.15
> [2018-06-19 18:07:52.678093] mircommon: Loading module: ./bin/../lib/server-modules/input-evdev.so.7
> [2018-06-19 18:07:52.679549] <DEBUG> mirserver: Using Linux VT subsystem for session management
> [2018-06-19 18:07:52.679584] mirplatform: Found graphics driver: mir:mesa-x11 (version 0.31.2)
> [2018-06-19 18:07:52.681919] mirplatform: Found graphics driver: mir:mesa-kms (version 0.31.2)
> [2018-06-19 18:07:52.681987] <DEBUG> eglstream: Found EGLDeviceEXT with device extensions: EGL_NV_device_cuda EGL_EXT_device_drm
> [2018-06-19 18:07:52.724791] mirplatform: Found graphics driver: mir:eglstream-kms (version 0.31.2)
> [2018-06-19 18:07:52.724940] mirserver: Selected driver: mir:mesa-kms (version 0.31.2)
> [2018-06-19 18:07:52.725100] <DEBUG> mirserver: Using Linux VT subsystem for session management
> ERROR: /home/gerry/dev/Canonical/mir/src/server/graphics/default_configuration.cpp(172): Throw in function mir::DefaultServerConfiguration::the_graphics_platform()::<lambda()>
> Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::runtime_error> >
> std::exception::what: Exception while creating graphics platform
> ERROR: /home/gerry/dev/Canonical/mir/src/server/console/linux_virtual_terminal.cpp(172): Throw in function {anonymous}::DRMDevice::DRMDevice(mir::VTFileOperations&, const char*, std::unique_ptr<mir::Device::Observer>, const std::shared_ptr<mir::LinuxVirtualTerminal::DeviceList>&)
> Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::system_error> >
> std::exception::what: Failed to claim DRM master: Invalid argument
> [boost::errinfo_file_name_*] = /dev/dri/card0

It isn't starting up. My bad?

**Comment by AlanGriffiths** [comment]
>  @gerboland "sudo XDG_RUNTIME_DIR=/tmp ./bin/miral-shell --vt 1"

Two things occur to me:
1. doesn't gdm already hold DRM Master on VT1? Does a different VT work?
2. how about switching to a VT and then using miral-shell without "--vt" (thereby using Logind, not LinuxVurtualTerminal)?

**Comment by gerboland** [comment]
Rebasing this on top of master, this error goes away. I'm testing on my NVidia machine, and Mir is still often choosing mesa-kms:
- if launching via a logged-in VT (but not specifying a VT), mir doesn't consider the eglstream-kms platform, uses mesa-kms and fails with "libEGL warning: DRI2: failed to create dri screen". It does use logind.
- If I launch exactly the same command, but from SSH, mir still uses mesa-kms, prints that libEGL error, but it does run and render (using llvmpipe).
- if I try so specify a different VT to the one logged in, Mir fails with 

`[2018-07-18 10:08:00.202269] mirserver: Selected driver: mir:eglstream-kms (version 0.32.1)
[2018-07-18 10:08:00.202414] <DEBUG> mirserver: Using Linux VT subsystem for session management
ERROR: /home/gerry/dev/Canonical/mir/src/platforms/eglstream-kms/server/egl_output.cpp(281): Throw in function void mir::graphics::eglstream::kms::EGLOutput::configure(size_t)
Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::runtime_error> >
std::exception::what: Failed to find EGLOutputEXT corresponding to DRM CRTC`


Doesn't matter if via VT or SSH.

**Comment by gerboland** [comment]
Oh, and manually specifying the graphics platform, via SSH:

`sudo XDG_RUNTIME_DIR=/tmp  ./bin/miral-shell --platform-graphics-lib=lib/server-modules/graphics-eglstream-kms.so.15
MIR_CLIENT_PLATFORM_PATH=./bin/../lib/client-modules/
MIR_SERVER_PLATFORM_PATH=./bin/../lib/server-modules/
LD_LIBRARY_PATH=./bin/../lib
exec=./bin/miral-shell.bin
[2018-07-18 10:39:31.393549] mirserver: Starting
[2018-07-18 10:39:31.393785] mirserver: Selected driver: mir:eglstream-kms (version 0.32.1)
[2018-07-18 10:39:31.395914] <DEBUG> mirserver: Not using logind for session management: /home/gerry/dev/Canonical/mir/src/server/console/logind_console_services.cpp(182): Throw in function std::__cxx11::string {anonymous}::object_path_for_current_session(LogindSeat*)
Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::runtime_error> >
std::exception::what: Seat has no active session

[2018-07-18 10:39:31.396188] <DEBUG> mirserver: Using Linux VT subsystem for session management
ERROR: /home/gerry/dev/Canonical/mir/src/platforms/eglstream-kms/server/egl_output.cpp(281): Throw in function void mir::graphics::eglstream::kms::EGLOutput::configure(size_t)
Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::runtime_error> >
std::exception::what: Failed to find EGLOutputEXT corresponding to DRM CRTC`

and in a VT, it crashes after printing "Using logind for session management"

**Comment by RAOF** [comment]
OK, that's the same thing that I see locally, which means that it's not my laptop being all weird and hybrid-y.

Bah. Looks like something's changed in the NVIDIA driver?

File: src/platforms/common/server/kms-utils/drm_mode_resources.cpp [hunk 1]
```diff
@@ -318,7 +318,7 @@ auto mgk::DRMModeResources::encoders() const -> detail::ObjectCollection<DRMMode
 
 auto mgk::DRMModeResources::crtcs() const -> detail::ObjectCollection<DRMModeCrtcUPtr, &get_crtc>
 {
-    return detail::ObjectCollection<DRMModeCrtcUPtr, &get_crtc>{drm_fd, resources->crtcs, resources->encoders + resources->count_crtcs};
+    return detail::ObjectCollection<DRMModeCrtcUPtr, &get_crtc>{drm_fd, resources->crtcs, resources->crtcs + resources->count_crtcs};
```
Comment (RAOF) [inline src/platforms/common/server/kms-utils/drm_mode_resources.cpp]: Endless screaming

**Comment by RAOF** [comment]
@gerboland Let's try *this* ping :grinning: 

This successfully brings up a server on my machine.

**Comment by gerboland** [comment]
It works on my nvidia box. mir_demo_shell shows be black screen and cursor, and uses eglstream-kms and proper GL. I can run glmark2-wayland and see stuff (it is using llvmpipe)

File: src/platforms/common/server/kms-utils/drm_mode_resources.cpp (hunk 1 above)
Comment (gerboland) [reply src/platforms/common/server/kms-utils/drm_mode_resources.cpp]: ouch

File: src/platforms/eglstream-kms/server/platform_symbols.cpp [hunk 2]
```diff
@@ -44,6 +45,10 @@ namespace mgc = mir::graphics::common;
 namespace mo = mir::options;
 namespace mge = mir::graphics::eglstream;
 
+namespace
+{
+int const EGL_DRM_MASTER_FD_EXT{0x333C};
```
Comment (gerboland) [inline src/platforms/eglstream-kms/server/platform_symbols.cpp]: Can you put this, adds context to the magic
```
/* XXX khronos eglext.h does not yet have EGL_DRM_MASTER_FD_EXT */
#if !defined(EGL_DRM_MASTER_FD_EXT)
#define EGL_DRM_MASTER_FD_EXT                   0x333C
#endif
```

Comment (AlanGriffiths) [reply src/platforms/eglstream-kms/server/platform_symbols.cpp]: There's already a similar incantation in src/platforms/eglstream-kms/server/platform.cpp, so I would view this as pre-existing.

However, this code will break on any platform that #defines EGL_DRM_MASTER_FD_EXT. We should prevent that with an incantation similar to that @gerboland suggests. (I'd be happy to write or approve a follow-up PR fixing both instances.)

Comment (gerboland) [reply src/platforms/eglstream-kms/server/platform_symbols.cpp]: Ok, will approve now then
bors r+

**Pull request merged**